tools/nixpkgs-hammer

```diff
@@ -112,6 +116,65 @@ def run_external_checks(
     return dict(result)
 
 
+def run_fetchurl_checks(
+    fetchurl_calls: List[Dict[str, Any]], excluded_rules: List[str]
+) -> Dict[str, Report]:
+
+    if "out-of-repo-commit" in excluded_rules:
+        return {}
+
+    github_fetch = namedtuple("github_fetch", ["url", "owner", "repo", "sha"])
+
+    def github_commit_fetches() -> Iterable[github_fetch]:
+        for call in fetchurl_calls:
+            url = call.get("url", "")
+            if not isinstance(url, str):
+                continue
+            m1 = re.search(
+                r"github\.com/(.*)/(.*)/(?:archive|commit|pull/\d+/commits)/(.{40})\.",
+                url,
+            )
+            if m1 is not None:
+                yield (url,) + m1.groups()
+
+    def _fetch_data(url: str, owner: str, repo: str, sha: str) -> Tuple[str, str]:
+        branch_commit_url = f"https://github.com/{owner}/{repo}/branch_commits/{sha}"
+        try:
+            resp = urllib.request.urlopen(branch_commit_url, timeout=20)
+            if resp.status != 200:
+                return None
+            return (url, resp.read())
+        except urllib.error.HTTPError:
+            return None
+
+    github_responses = []
+    with concurrent.futures.ThreadPoolExecutor(max_workers=10) as executor:
+        future_to_url = (
+            executor.submit(_fetch_data, *value) for value in github_commit_fetches()
+        )
+        for future in concurrent.futures.as_completed(future_to_url):
+            github_responses.append(future.result())
+
+    reports = []
+    for url, response in github_responses:
+        if b"This commit does not belong to any branch on this repository." in response:
+            reports.append(
+                Report(
+                    name="out-of-repo-commit",
+                    msg=textwrap.dedent(
+                        f"""
+                            The dependency chain for these expressions includes {url},
+                            which references a commit which does not belong to any
+                            branch on this repository, and is in fact associated with
+                            a fork of this repository.
+                        """
+                    ).strip(),
+                    locations=[],
+                )
+            )
+    return {"<unknown>": reports}
```

Comment:
Here's an example of what it looks like:

```
nix run -f ~/projects/nixpkgs-hammering -c nixpkgs-hammer -f . openexr -e attribute-ordering -e no-build-output -e unused-argument -e maintainers-missing
When evaluating attribute ‘openexr’:
No issues found.

When evaluating attribute ‘<unknown>’:
warning: out-of-repo-commit
The dependency chain for these expressions includes https://github.com/AcademySoftwareFoundation/openexr/commit/6442fb71a86c09fb0a8118b6dbd93bcec4883a3c.patch,
which references a commit which does not belong to any
branch on this repository, and is in fact associated with
a fork of this repository.
See: https://github.com/jtojnar/nixpkgs-hammering/blob/master/explanations/out-of-repo-commit.md
```

Currently, the technical limitation is that I don't know how to associate the fetchurl call with the attr -- that is, if you're simultaneously running nixpkgs-hammering on two or more attrs and the dependency chain for one of the attrs triggers a report of this type, I don't know how to determine which attr triggered the report.

Reply:
Perhaps we could print the `outPath` of the `fetchurl` call, get a drv file from that and then check which of the attributes has that in closure.

Reply:
Yes, that seems reasonable. As best I can tell, we can't compute the closure from within the nix language, and instead that needs to be done from the cmdline (`nix why-depends`, `nix show-derivation -r` or `nix-store --query --graphml`, or similar). Is that also your understanding?

Reply:
With IFD, you might run those commands using `runCommand` or write a new store-querying builting. But yeah, I am not aware of any current facility to query Nix store within Nix language. (Well, you could compute the closure by literally reading the `*.drv` store paths but letting Nix get the values from the SQLite database will probably be much faster.)

We will probably want to print the why-depends result in the check output anyway, to make it clear from which derivation it comes from.